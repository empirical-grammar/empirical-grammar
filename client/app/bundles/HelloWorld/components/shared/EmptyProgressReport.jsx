import React from 'react';

export default React.createClass({

  render() {
    let link,
      buttonText,
      content;
    if (this.props.missing === 'activities') {
      link = '/teachers/classrooms/assign_activities';
      buttonText = 'Assign an Activity';
      content = 'In order to access our different reports, you need to assign activities to your students.';
    } else if (this.props.missing === 'students') {
      link = '/teachers/classrooms/invite_students';
      buttonText = 'Invite Students';
      content = 'In order to access our different reports, you need to invite your students and assign activities.'
    } else {
      link = '/teachers/classrooms/new';
      buttonText = 'Create a Class';
      content = 'In order to access our different reports, you need to create a class and assign activities to your students.'
    }
 		return (
   <div className="empty-progress-report">
     <img src='/images/empty_state_illustration.png' />
     <h1>You have no reports yet!</h1>
     <p>{content}</p>
     <button onClick={() => { window.location = link; }} className="button-green create-unit featured-button">{buttonText}</button>
     <a href="/teacher_resources">Teacher Resources</a>
   </div>
 );
  },
});
