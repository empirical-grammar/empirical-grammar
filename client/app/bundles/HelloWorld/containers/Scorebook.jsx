import React from 'react';
import Scrollify from '../components/modules/scrollify';
import $ from 'jquery';
import request from 'request';
import _ from 'underscore';
import TableFilterMixin from '../components/general_components/table/sortable_table/table_filter_mixin.js';
import StudentScores from '../components/scorebook/student_scores';
import LoadingIndicator from '../components/shared/loading_indicator';
import ScorebookFilters from '../components/scorebook/scorebook_filters';
import ScoreLegend from '../components/scorebook/score_legend';
import AppLegend from '../components/scorebook/app_legend.jsx';
import EmptyProgressReport from '../components/shared/EmptyProgressReport';

export default React.createClass({
  mixins: [TableFilterMixin],

  getInitialState() {
    this.modules = {
      scrollify: new Scrollify(),
    };
    return {
      units: [],
      classrooms: this.props.allClassrooms,
      selectedUnit: {
        name: 'All Activity Packs',
        value: '',
      },
      selectedClassroom: this.props.selectedClassroom,
      classroomFilters: this.props.allClassrooms,
      unitFilters: [],
      scores: new Map(),
      beginDate: null,
      premium_state: this.props.premium_state,
      endDate: null,
      currentPage: 0,
      loading: false,
      is_last_page: false,
      noLoadHasEverOccurredYet: true,
    };
  },

  componentDidMount() {
    this.fetchData();
    this.getUpdatedUnits(this.props.selectedClassroom.value);
    this.modules.scrollify.scrollify('#page-content-wrapper', this);
  },

  formatDate(date) {
    if (date) {
      return `${date.year()}-${date.month() + 1}-${date.date()}`;
    }
  },

  fetchData() {
    const newCurrentPage = this.state.currentPage + 1;
    this.setState({ loading: true, currentPage: newCurrentPage, });
    $.ajax({
      url: '/teachers/classrooms/scores',
      data: {
        current_page: newCurrentPage,
        classroom_id: this.state.selectedClassroom.value,
        unit_id: this.state.selectedUnit.value,
        begin_date: this.formatDate(this.state.beginDate),
        end_date: this.formatDate(this.state.endDate),
        no_load_has_ever_occurred_yet: this.state.noLoadHasEverOccurredYet,
      },
      success: this.displayData,
    });
  },

  getUpdatedUnits(classroomId) {
    const loadingUnit = { name: 'Loading...', value: '', };
    this.setState({ unitFilters: [loadingUnit], selectedUnit: loadingUnit, });
    const that = this;
    request.get({
      url: `${process.env.DEFAULT_URL}/teachers/classrooms/${classroomId}/units`,
    }, (error, httpStatus, body) => {
      const parsedBody = JSON.parse(body);
      that.setState({ unitFilters: parsedBody.units, selectedUnit: { name: 'All Activity Packs', value: '', }, });
    });
  },

  checkMissing(newScores) {
    if (!this.state.classroomFilters || this.state.classroomFilters.length === 0) {
      return 'classrooms';
    } else if (!newScores || newScores.size === 0) {
      return 'activities';
    }
  },

  displayData(data) {
    this.setState({
      classroomFilters: this.props.allClassrooms,
      is_last_page: data.is_last_page,
      premium_state: data.premium_state,
      noLoadHasEverOccurredYet: false,
    });
    const newScores = new Map(this.state.scores);
    data.scores.forEach((s) => {
      // add the score to the user scores arr or create a new one
      newScores.has(s.user_id) || newScores.set(s.user_id, { name: s.name, scores: [], });
      const scores = newScores.get(s.user_id).scores;
      scores.push({
        caId: s.ca_id,
        // activitySessionId: s.id,
        userId: s.user_id,
        updated: s.updated_at,
        name: s.activity_name,
        percentage: s.percentage,
        activity_classification_id: s.activity_classification_id, });
    });
    this.setState({ loading: false, scores: newScores, missing: this.checkMissing(newScores), });
  },

  selectUnit(option) {
    this.setState({
      scores: new Map(),
      currentPage: 0,
      selectedUnit: option,
    }, this.fetchData);
  },

  selectClassroom(option) {
    this.getUpdatedUnits(option.value);
    this.setState({
      scores: new Map(),
      currentPage: 0,
      selectedClassroom: option,
    }, this.fetchData
  );
  },

  selectDates(val1, val2) {
    this.setState({
      scores: new Map(),
      currentPage: 0,
      beginDate: val1,
      endDate: val2,
    }, this.fetchData);
  },

  render() {
    let content,
      loadingIndicator;
    const scores = [];
    let index = 0;
    this.state.scores.forEach((s) => {
      index += 0;
      const sData = s.scores[0];
      scores.push(<StudentScores key={`${sData.userId}`} data={{ scores: s.scores, name: s.name, activity_name: sData.activity_name, userId: sData.userId, }} premium_state={this.props.premium_state} />);
    });
    if (this.state.loading) {
      loadingIndicator = <LoadingIndicator />;
    } else {
      loadingIndicator = null;
    }

    if (this.state.missing) {
      content = <EmptyProgressReport missing={this.state.missing} />;
    } else {
      content =
        (<div>
          {scores}
          {loadingIndicator}
        </div>);
    }
    return (
      <div className="page-content-wrapper">
        <div className="tab-pane" id="scorebook">
          <div className="container">
            <section className="section-content-wrapper">
              <ScorebookFilters
                selectedClassroom={this.state.selectedClassroom}
                classroomFilters={this.state.classroomFilters}
                selectClassroom={this.selectClassroom}
                selectedUnit={this.state.selectedUnit}
                unitFilters={this.state.unitFilters}
                selectUnit={this.selectUnit} selectDates={this.selectDates}
              />
              <ScoreLegend />
              <AppLegend />
            </section>
          </div>
          {content}
        </div>
      </div>
    );
  },
});
