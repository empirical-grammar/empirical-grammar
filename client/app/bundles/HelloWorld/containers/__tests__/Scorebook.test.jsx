import React from 'react';
import { shallow, mount } from 'enzyme';

import moment from 'moment';

import Scorebook from '../Scorebook.jsx';
import EmptyProgressReport from '../../components/shared/EmptyProgressReport';
import LoadingIndicator from '../../components/shared/loading_indicator';
import StudentScores from '../../components/scorebook/student_scores';
import ScorebookFilters from '../../components/scorebook/scorebook_filters';
import ScoreLegend from '../../components/scorebook/score_legend';
import AppLegend from '../../components/scorebook/app_legend.jsx';

const resolvedScores = new Map();
resolvedScores.set('441555', { name: 'blah blah', scores: [{ caId: '341930', userId: '441555', updated: '2016-09-16 15:39:00.775325', name: 'America Used to be a Different Place', percentage: '1', activity_classification_id: '1', }], }, );
const classrooms = [{ name: 'A', id: 1, }, { name: 'B', id: 2, }, { name: 'C', id: 3, }];
const units = [{ name: 'Something', id: 4, }, { name: 'Someone', id: 5, }, { name: 'Somewhere', id: 6, }];
const rawScores = [{ user_id: '441555', ca_id: '341930', name: 'blah blah', activity_classification_id: '1', activity_name: 'America Used to be a Different Place', updated_at: '2016-09-16 15:39:00.775325', percentage: '1', id: '5951806', }];

const data = {
  teacher: { premium_state: 'trial', },
  is_last_page: false,
  classrooms,
  units,
  scores: JSON.parse(JSON.stringify(rawScores)),
};

describe('Scorebook component', () => {
  it('should render', () => {
    const wrapper = shallow(<Scorebook />);
    expect(wrapper).toMatchSnapshot();
  });
  //
  describe('if state.missing has a value', () => {
    const wrapper = shallow(<Scorebook />);
    wrapper.setState({ missing: 'activities', });

    it('renders EmptyProgressReport', () => {
      expect(wrapper.find(EmptyProgressReport).length).toEqual(1);
    });

    it('does not render LoadingIndicator', () => {
      expect(wrapper.find(LoadingIndicator).length).toEqual(0);
    });

    it('still renders ScorebookFilters', () => {
      expect(wrapper.find(ScorebookFilters).length).toEqual(1);
    });

    it('still renders ScoreLegend', () => {
      expect(wrapper.find(ScoreLegend).length).toEqual(1);
    });

    it('still renders AppLegend', () => {
      expect(wrapper.find(AppLegend).length).toEqual(1);
    });

    it('does not render StudentScores', () => {
      expect(wrapper.find(StudentScores).length).toEqual(0);
    });

    // it('does not get rendered if it is not passed a value for missing', () => {
    //   const wrapper = shallow(<Scorebook missing={''}/>)
    //   expect(wrapper.find(EmptyProgressReport).length).toEqual(0)
    // })
  });

  describe('if state.loading is true', () => {
    const wrapper = shallow(<Scorebook />);
    wrapper.setState({ loading: true, });

    it('renders LoadingIndicator', () => {
      expect(wrapper.find(LoadingIndicator).length).toEqual(1);
    });

    it('renders ScorebookFilters', () => {
      expect(wrapper.find(ScorebookFilters).length).toEqual(1);
    });

    it('renders ScoreLegend', () => {
      expect(wrapper.find(ScoreLegend).length).toEqual(1);
    });

    it('renders AppLegend', () => {
      expect(wrapper.find(AppLegend).length).toEqual(1);
    });

    it('does not render StudentScores', () => {
      expect(wrapper.find(StudentScores).length).toEqual(0);
    });
  });

  describe('if it gets data when it had none', () => {
    const wrapper = shallow(<Scorebook />);
    wrapper.setState({ currentPage: 1, scores: new Map(), });
    wrapper.instance().displayData(data);

    it('does not render LoadingIndicator', () => {
      expect(wrapper.find(LoadingIndicator).length).toEqual(0);
    });

    it('renders ScorebookFilters', () => {
      expect(wrapper.find(ScorebookFilters).length).toEqual(1);
    });

    it('renders ScoreLegend', () => {
      expect(wrapper.find(ScoreLegend).length).toEqual(1);
    });

    it('renders AppLegend', () => {
      expect(wrapper.find(AppLegend).length).toEqual(1);
    });

    it('updates the state.scores to be equal to the new ones', () => {
      expect(wrapper.state('scores')).toEqual(resolvedScores);
    });
  });

  describe('if it gets data when it had some', () => {
    const newResScores = new Map(resolvedScores);
    const newRawScores = Array.slice(rawScores);
    newRawScores[0].user_id = 'fakey';
    const wrapper = shallow(<Scorebook allClassrooms={classrooms} />);
    data.scores = newRawScores;
    wrapper.setState({ scores: newResScores, classroomFilters: classrooms, });
    wrapper.instance().displayData(data);
    it('does not render LoadingIndicator', () => {
      expect(wrapper.find(LoadingIndicator).length).toEqual(0);
    });

    it('renders ScorebookFilters', () => {
      expect(wrapper.find(ScorebookFilters).length).toEqual(1);
    });

    it('renders ScoreLegend', () => {
      expect(wrapper.find(ScoreLegend).length).toEqual(1);
    });

    it('renders AppLegend', () => {
      expect(wrapper.find(AppLegend).length).toEqual(1);
    });

    it('updates the state.scores to add the new ones', () => {
      const updated = { 441555: { name: 'blah blah', scores: [{ activity_classification_id: '1', caId: '341930', name: 'America Used to be a Different Place', percentage: '1', updated: '2016-09-16 15:39:00.775325', userId: '441555', }], }, fakey: { name: 'blah blah', scores: [{ activity_classification_id: '1', caId: '341930', name: 'America Used to be a Different Place', percentage: '1', updated: '2016-09-16 15:39:00.775325', userId: 'fakey', }], }, };
      const newScores = new Map(resolvedScores);
      newScores.set('441555', { name: 'blah blah', scores: [{ activity_classification_id: '1', caId: '341930', name: 'America Used to be a Different Place', percentage: '1', updated: '2016-09-16 15:39:00.775325', userId: '441555', }], });
      newScores.set('fakey', { name: 'blah blah', scores: [{ activity_classification_id: '1', caId: '341930', name: 'America Used to be a Different Place', percentage: '1', updated: '2016-09-16 15:39:00.775325', userId: 'fakey', }], });
      expect(wrapper.state('scores')).toEqual(newScores);
    });

    it('renders as many StudentScores as there are state.scores', () => {
      expect(wrapper.find(StudentScores).length).toEqual(wrapper.state('scores').size);
    });
  });
});
