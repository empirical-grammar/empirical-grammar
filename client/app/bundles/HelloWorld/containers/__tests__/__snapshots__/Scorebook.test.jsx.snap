// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scorebook component should render 1`] = `
<div
  className="page-content-wrapper"
>
  <div
    className="tab-pane"
    id="scorebook"
  >
    <div
      className="container"
    >
      <section
        className="section-content-wrapper"
      >
        <scorebook_filters
          selectClassroom={[Function]}
          selectDates={[Function]}
          selectUnit={[Function]}
          selectedUnit={
            Object {
              "name": "All Activity Packs",
              "value": "",
            }
          }
          unitFilters={Array []}
        />
        <score_legend />
        <app_legend />
      </section>
    </div>
    <div />
  </div>
</div>
`;
